src/library_wasi.js

```diff
@@ -261,17 +260,41 @@ var WasiLibrary = {
     if (printCharBuffers[1].length) printChar(1, {{{ charCode("\n") }}});
     if (printCharBuffers[2].length) printChar(2, {{{ charCode("\n") }}});
   },
-  fd_write__deps: ['$flush_NO_FILESYSTEM', '$printChar'],
-  fd_write__postset: () => addAtExit('flush_NO_FILESYSTEM()'),
+#if SYSCALLS_REQUIRE_FILESYSTEM
+  fd_write__postset: () => addAtExit('if(ENVIRONMENT_IS_WASM_WORKER) flush_NO_FILESYSTEM()'),
 #else
-  fd_write__deps: ['$printChar'],
+  fd_write__postset: () => addAtExit('flush_NO_FILESYSTEM()'),
+#endif
+#endif
+
+  fd_write__deps: [
+#if SYSCALLS_REQUIRE_FILESYSTEM
+    '$doWritev',
+#endif
+#if (!SYSCALLS_REQUIRE_FILESYSTEM || WASM_WORKERS)
+    '$printChar',
+#if (!MINIMAL_RUNTIME || EXIT_RUNTIME)
+    '$flush_NO_FILESYSTEM',
 #endif
+#endif
+],
   fd_write: (fd, iov, iovcnt, pnum) => {
 #if SYSCALLS_REQUIRE_FILESYSTEM
+#if WASM_WORKERS
+  if (!ENVIRONMENT_IS_WASM_WORKER) {
+#endif // WASM_WORKERS
     var stream = SYSCALLS.getStreamFromFD(fd);
     var num = doWritev(stream, iov, iovcnt);
-#else
-    // hack to support printf in SYSCALLS_REQUIRE_FILESYSTEM=0
+#if WASM_WORKERS
+  } else {
+    // Hack to support printf in WASM workers in a simple way.
+    // We support at least stdout, stderr.
+#endif // WASM_WORKERS
+#endif // SYSCALLS_REQUIRE_FILESYSTEM
+#if !SYSCALLS_REQUIRE_FILESYSTEM || WASM_WORKERS
+#if !SYSCALLS_REQUIRE_FILESYSTEM
+    // Hack to support printf in SYSCALLS_REQUIRE_FILESYSTEM=0
+#endif
```

Comment:
Kind of shame that we now have to support the hybrid approach of both real FS stuff and fake FS stuff in the same build.  

To avoid some of these confusing `#if` maybe we could split out the simple/dummy version fs_write and give is its own name  (e.g. `fdWriteNOFS`)

Then we could generate something like:

```
fd_write: (fd, iov, iovcnt, pnum) => {
   if (ENVIRONMENT_IS_WASM_WORKER) {
      return fdWriteNOFS(...)
   }
   ... nomral FS code ..
}
```

And in the non-worker case where no filesystem is needed you could use an alias:

```
#if !SYSCALLS_REQUIRE_FILESYSTEM
fd_write: '$fdWriteNOFS',`
#endif
```

Reply:
Modified as required.